File: counting.sort.py
def counting_sort(arr):
    # Găsește valoarea maximă din array
    max_val = max(arr)
    min_val = min(arr)
    range_of_elements = max_val - min_val + 1

    # Creează un array de numărare (count array) și îl inițializează cu 0
    count = [0] * range_of_elements

    # Array pentru rezultatul sortării
    output = [0] * len(arr)

    # Contorizează frecvența fiecărui element
    for num in arr:
        count[num - min_val] += 1

    # Modifică count astfel încât să conțină pozițiile finale ale elementelor
    for i in range(1, len(count)):
        count[i] += count[i - 1]

    # Construiește array-ul sortat
    for num in reversed(arr):
        output[count[num - min_val] - 1] = num
        count[num - min_val] -= 1

    return output

# Exemplu 
if __name__ == "__main__":
    arr = [4, 2, 2, 8, 3, 3, 1]
    sorted_arr = counting_sort(arr)
    print("Array sortat:", sorted_arr)

File: Counting_Sort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Funcție pentru a realiza sortarea Counting Sort
void countingSort(vector<int>& arr) {
    // Verificăm dacă array-ul este gol
    if (arr.empty()) return;

    // Pas 1: Găsim valoarea maximă din array pentru a determina dimensiunea count array-ului
    int maxVal = *max_element(arr.begin(), arr.end());

    // Pas 2: Inițializăm un array count de dimensiune maxVal + 1 (deoarece indexările încep de la 0)
    vector<int> count(maxVal + 1, 0);

    // Pas 3: Numărăm aparițiile fiecărui element
    for (int num : arr) {
        count[num]++;
    }

    // Pas 4: Actualizăm array-ul original în funcție de count
    int index = 0; // Index pentru a salva în array-ul original
    for (int i = 0; i <= maxVal; i++) {
        // Dacă count[i] > 0, repetăm i de count[i] ori
        while (count[i] > 0) {
            arr[index++] = i; // Salvăm elementul i în array-ul original
            count[i]--; // Decrementăm count
        }
    }
}

// Funcție pentru a afișa un array
void printArray(const vector<int>& arr) {
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
}

int main() {
    // Definim un array de numere întregi
    vector<int> arr = { 4, 2, 2, 8, 3, 3, 1 };

    cout << "Array-ul original: ";
    printArray(arr);

    // Apelăm funcția de sortare
    countingSort(arr);

    cout << "Array-ul sortat: ";
    printArray(arr);
    return 0;
}